update codebase to 1.0 and 1.x (continue #27)

johnnychen94 opened the pull request:
This PR continues #27 and the road I choose is to:

- [x] fix those easy-to-fix broken tests in julia 1.0
- [x] fix type instability
- [x] update codes to 1.3 while keep backward compatibility to 1.0
- [ ] fix the random failure in `tst_scale.jl`
- [ ] support `OffsetArray 1.0`

There're some type-instability cases detected, which I've added a TODO comment for that, and probably fix them in a separate future PR.

To easily review the differences, the current base is set to `update0.7`, will change it to `master` before merging.

johnnychen94 committed:
add Project.toml

johnnychen94 committed:
update test for cache and minor fix

johnnychen94 committed:
update test for rotation and minor fix

johnnychen94 committed:
update flipdm to reverse

https://github.com/JuliaLang/julia/pull/26369

johnnychen94 committed:
update test for crop and minor fix

johnnychen94 committed:
update test for resize

johnnychen94 committed:
update test for scale and minor fix

johnnychen94 committed:
update test for zoom and minor fix

johnnychen94 committed:
WIP: update test for distortion and minor fix

johnnychen94 committed:
update test for either and minor fix

johnnychen94 committed:
fix pipeline test

johnnychen94 committed:
update summary reference for distortedview

johnnychen94 committed:
WIP: fix tests for augment and operations

johnnychen94 committed:
drop julia 0.7 and update CI

johnnychen94 commented on src/distortedview.jl:
```diff
@@ -22,11 +32,11 @@ function Base.showarg(io::IO, A::DistortedView, toplevel)
 
 # inline speeds up ~30%
-@inline function Base.getindex(A::DistortedView, i::Int, j::Int)
+function Base.getindex(A::DistortedView, i::Int, j::Int)
```
```suggestion
@inline function Base.getindex(A::DistortedView, i::Int, j::Int)
```

johnnychen94 commented on src/distortedview.jl:
```diff
@@ -12,8 +12,18 @@ struct DistortedView{T,P<:AbstractMatrix,E<:AbstractExtrapolation,G,D} <: Abstra
+end
+function DistortedView(A::InvWarpedView, grid::AbstractArray{Float64, 3})
+    DistortedView(collect(A), grid)
+end
```
These lines are used to fix the test cases, but I believe that there're better solutions

johnnychen94 commented on src/operations/resize.jl:
```diff
@@ -79,8 +79,9 @@ function padrange(range::AbstractUnitRange, pad)
-    Rin, Rout = CartesianRange(axes(img)), CartesianRange(op.size)
-    sf = map(/, (last(Rout)-first(Rout)+1).I, (last(Rin)-first(Rin)+1).I)
+    Rin, Rout = CartesianIndices(axes(img)), CartesianIndices(op.size)
+    @assert first(Rout) == CartesianIndex(1, 1)
```
```suggestion
```

johnnychen94 committed:
use FileIO v1.1.0

https://github.com/JuliaIO/FileIO.jl/issues/245

johnnychen94 committed:
add 1.x tests to CI

johnnychen94 committed:
revert back unnecessary changes

johnnychen94 committed:
test windows in travis

johnnychen94 commented:
@Evizero could you enable Auto Cancellation of the previous build in travis?

Evizero commented:
> @Evizero could you enable Auto Cancellation of the previous build in travis?

It looks enabled already

<img width="838" alt="grafik" src="https://user-images.githubusercontent.com/10854026/70888208-109bc200-1fe0-11ea-922d-867d19c0b06e.png">

johnnychen94 committed:
fix randomly failed test cases for inacuraccy reasons

johnnychen94 committed:
update to julia 1.1

* maybe_copy doesn't explain its usage
* replace Slice by OffsetArrays.IdentityUnitRange

https://github.com/JuliaArrays/OffsetArrays.jl/pull/62

johnnychen94 committed:
update to julia 1.2 -- part I

johnnychen94 committed:
WIP: update to julia 1.2 -- part II

detect some type instable test cases, only the outmost wrapper is
instable here

johnnychen94 committed:
unify how types are displayed

Only meta operations such as Either, Pipeline doesn't have Agumentor
prefix

johnnychen94 committed:
add more test versions

johnnychen94 commented on src/utils.jl:
```diff
@@ -32,13 +32,21 @@ function safe_rand(args...)
+Data copy only happens when necessary. For example, views returned by `view`,
+`permuteddimsview` are such cases.
+"""
+@inline contiguous(A::OffsetArray) = A
```
I did a rename here to make it more self-explained

johnnychen94 commented on test/operations/tst_either.jl:
```diff
@@ -117,68 +117,68 @@ end
             @test Augmentor.supports_permute(op) === false
-            @test @inferred(Augmentor.applyeager(op, img)) == rotl90(rect)
-            @test typeof(Augmentor.applyeager(op, img)) <: OffsetArray
+            @test @inferred(Augmentor.applyeager(op, img, Augmentor.randparam(op, img))) == rotl90(rect) # TODO: update 1.2: type instable if remove Augmentor.randparam
```
Type instability is detected here for Julia 1.2; the lucky part is that only the out-most wrapper is affected.

Evizero replied:
Is it `randparam` then?

johnnychen94 replied:
Sounds reasonable, I'll check that; not very familiar with this kind of issue at present.

johnnychen94 replied:
It turns out that the following part doesn't work anymore

https://github.com/Evizero/Augmentor.jl/blob/6055db7493352257ddc7b97d63ecaf8ccb43faea/src/operations/either.jl#L175-L181

a typical error is like
```txt
ERROR: return type OffsetArray{Gray{Normed{UInt8,8}},2,Array{Gray{Normed{UInt8,8}},2}} does not match inferred return type OffsetArray
```

I need time to investigate it, but for now, I'll pause it for a few days just to use this package/branch to catch up with my daily work.

Evizero commented on src/operations/resize.jl:
```diff
@@ -79,8 +79,8 @@ function padrange(range::AbstractUnitRange, pad)
-    Rin, Rout = CartesianRange(axes(img)), CartesianRange(op.size)
-    sf = map(/, (last(Rout)-first(Rout)+1).I, (last(Rin)-first(Rin)+1).I)
+    Rin, Rout = CartesianIndices(axes(img)), CartesianIndices(op.size)
+    sf = Tuple(last(Rout)) ./ Tuple(last(Rin) - first(Rin) + first(Rout))
```
This doesn't quite look equivalent on first glance. is it?

johnnychen94 replied:
It turns out that `CartesianIndices(::Tuple)` returns an index array whose first element is `CartesianIndex(1, 1)`. But yes,  your're right, what we need here is to substitute the 1-base `CartesianIndex(1, 1)` and it has nothing to do with `Rout`.

johnnychen94 replied:
fixed in https://github.com/Evizero/Augmentor.jl/pull/31/commits/67e97846c1025492cad36196e28fa6d5945d1175

Evizero commented on test/tst_distortedview.jl:
```diff
@@ -116,16 +116,16 @@ end
     @test size(dv2) == size(camera)
     @test eltype(dv2) == eltype(camera)
-    @test summary(dv2) == "512×512 Augmentor.DistortedView(::OffsetArray{Gray{N0f8},2}, ::Array{Float64,3} as 3×3 vector field) with element type ColorTypes.Gray{FixedPointNumbers.Normed{UInt8,8}}"
+    @test summary(dv2) == "512×512 Augmentor.DistortedView(::Array{Gray{N0f8},2}, ::Array{Float64,3} as 3×3 vector field) with eltype Gray{Normed{UInt8,8}}"
```
What happened here?

johnnychen94 replied:
Looks like it's related to my previous added specification to `DistortedView`

It's a WIP commit, I'll recheck that later when I fix julia 1.3 deprecations.

johnnychen94 replied:
Rechecked in https://github.com/Evizero/Augmentor.jl/commit/733e32bebf8176bc0843b9679029d3ae147047b9 and it should be a safe change here (as far as I understand the codebase)

To successfully compare two `DistortedView`s `A` and `B`, their axes should be the same, i.e., `axes(A)==axes(B)`, so I add the `plain_axes` in the construction method to make sure this. This is still lazy-mode since `plain_axes` doesn't copy data. The test in `tst_distortion.jl` checks that as well.

---

I tried to use git blame to find out how in the previous codes two arrays of different axes can compare successfully but I failed, so I find another path here.

Although I think this change meets the goal of `DistortedView`, it does change the behavior, hence you might need to review this commit carefully. 😂

Evizero reviewed (commented):
I reviewed all your code changes so far
Thanks for working on this

johnnychen94 committed:
move safe_rand to compat.jl

rand is thread-safe in julia 1.3

johnnychen94 committed:
don't allow failures for julia 1.3

johnnychen94 commented:
Good news is that except a random failure due to accuracy in `tst_scale.jl`, the whole codebase is now 1.3-compatible. Later commits are to continue WIP commits and address issues in review comments.

johnnychen94 committed:
fix commit "WIP: update 1.0"

johnnychen94 committed:
use explicit and intuitive CartesianIndex(1, 1)

Evizero commented:
amazing news!

johnnychen94 commented on src/utils.jl:
```diff
@@ -15,22 +15,6 @@ function use_testpattern()
-const rand_mutex = Ref{Threads.Mutex}()
-
-# constant overhead of about 80 ns compared to unsafe rand
-function safe_rand(args...)
```
I'm a little worried here that I might misunderstand what "threadsafe" means here. Is that mean the following two scripts should output the same results?

```julia
Random.seed!(0)
Threads.@threads for i in 1:5
    println(rand())
end
```

```julia
Random.seed!(0)
for i in 1:5
    println(rand())
end
```

Evizero replied:
no. This has no influence on the ordering that threads invoke `save_rand`. It really just is about the function being threadsafe so that consecutive return values (regardless of calling thread) can be considered random

johnnychen94 committed:
add method specialization for tweight

johnnychen94 committed:
allow FileIO 1.2

johnnychen94 committed:
fix type instability for Either

johnnychen94 committed:
update reference for FixedPointNumbers v0.7

johnnychen94 commented on src/compat.jl:
```diff
@@ -19,4 +21,10 @@ else
+# PR: https://github.com/JuliaImages/ImageTransformations.jl/pull/89
+if Pkg.installed()["ImageTransformations"] <= v"0.8.3"
+    Interpolations.tweight(A::AbstractArray{C}) where C<:FixedPoint = C
+end
```
This seems not a good idea. 👎 and it breaks the nightly test

johnnychen94 commented on src/operations/either.jl:
```diff
@@ -181,6 +181,14 @@ function applyeager(op::Either, img::AbstractArray, idx)
+end
+@inline function applyview(op::Either, img::AbstractArray)
+    applyview(op, img, randparam(op, img))
+end
```
The general definition is in `operation.jl`

```julia
for FUN in (:applyeager, :applylazy, :applypermute,
            :applyaffine, :applyaffineview,
            :applyaffine_common, :applyaffineview_common,
            :applyview, :applystepview)
    @eval begin
        function ($FUN)(op::Operation, imgs::Tuple)
            param = randparam(op, imgs)
            map(img -> ($FUN)(op, img, param), imgs)
        end
        @inline function ($FUN)(op::Operation, img::AbstractArray)
            ($FUN)(op, img, randparam(op, img))
        end
    end
end
```

But the inferred type for `Either` will be `Any`(without  using `_offsetarray`) or `OffsetArray`. Add a specialization here tells the compiler more type information.

johnnychen94 committed:
try: relax equality check for scale

johnnychen94 committed:
restore test cases in tst_scale.jl

johnnychen94 committed:
Revert "add method specialization for tweight"

This reverts commit a6983828a9a7ac581b6f314db8a31b7188c26656.

johnnychen94 committed:
Augmentor v0.6.0-pre

johnnychen94 commented:
This PR is growing longer and older with all these commits and it becomes hard to review. Hence I'm merging this PR regardless of two bugs not fixed( #38, #39)
